### PR TITLE
feat: adding switchable modes to LiDAR Visibility

### PR DESCRIPTION
**Pull request by badai-nguyen**
## Related Issue(required)
#357 

## Description(required)

This setup some options to exclude the effects of weak_first_return from objects:
1. Checking point clouds of object segments in both directions: azimuth and rings 
2. Adding 3 switchable modes of region of interest (ROI): 
- Free-space ROI
- Base_link based xyz-ROI
- LiDAR based azimuth-distance-ROI

## Review Procedure(required)

Switch and test 3 modes of ROI on a city road data.
Confirm the changing of visibility. 

## Related PR(optional)

https://github.com/tier4/aip_launcher/pull/21

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by miursh** [comment]
@badai-nguyen If you have made some change in launcher of this node, could you make another PR in launcher repository?

**Comment by badai-nguyen** [comment]
@miursh Sure. I will create PR in aip launcher repo.

File: sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp [hunk 1]
```diff
@@ -44,6 +53,11 @@ DualReturnOutlierFilterComponent::DualReturnOutlierFilterComponent(
     weak_first_local_noise_threshold_ =
       static_cast<int>(declare_parameter("weak_first_local_noise_threshold", 10));
     visibility_threshold_ = static_cast<float>(declare_parameter("visibility_threshold", 0.5));
+    ROI_mode_ = static_cast<uchar>(declare_parameter("ROI_mode", 0));
```
[the remark this answers is not in the document]
Comment (miursh) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp:56]: Parameter should not be magic number. Maybe words (string) is preferable.
If it is, user has to check the source code to run in desired mode. It is not user friendly.

File: sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp [hunk 2]
```diff
@@ -111,12 +142,23 @@ void DualReturnOutlierFilterComponent::filter(
     vertical_bins);  // TODO(davidw): this is for Pandar 40 only, make dynamic
   weak_first_pcl_input_ring_array.resize(vertical_bins);
 
+  const uint ring_number = vertical_bins;
+  const uint azimuth_steps = 36000;
+  float * pcl_input_distance_array = new float[ring_number * azimuth_steps];
+  float * pcl_weak_first_distance_array = new float[ring_number * azimuth_steps];
```
Comment (miursh) [inline sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp:148]: These variables start with "pcl" but I suppose they are not depends on PCL.
Remove pcl prefix.

**Comment by miursh** [comment]
@badai-nguyen Due to the source code change, https://github.com/autowarefoundation/autoware.universe/blob/tier4/proposal/sensing/pointcloud_preprocessor/docs/dual-return-outlier-filter.md needs to be modified.

File: sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp [hunk 3]
```diff
@@ -34,6 +34,15 @@ DualReturnOutlierFilterComponent::DualReturnOutlierFilterComponent(
...
   // set initial parameters
   {
+    x_max_ = static_cast<float>(declare_parameter("x_max", 18.0));
+    x_min_ = static_cast<float>(declare_parameter("x_min", -12.0));
+    y_max_ = static_cast<float>(declare_parameter("y_max", 2.0));
+    y_min_ = static_cast<float>(declare_parameter("y_min", -2.0));
+    z_max_ = static_cast<float>(declare_parameter("z_max", 10.0));
+    z_min_ = static_cast<float>(declare_parameter("z_min", 0.0));
+    min_azimuth_ = static_cast<float>(declare_parameter("min_azimuth", 13500.0));
+    max_azimuth_ = static_cast<float>(declare_parameter("max_azimuth", 22500.0));
```
[the remark this answers is not in the document]
Comment (badai-nguyen) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp:44]: Thank you for your comments, I will revise it.

[the remark this answers is not in the document]
Comment (badai-nguyen) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp]: 1. The "_deg" post-fix has been added
2. The variable range has been changed as physical quantities ( 0.0 ~ 360.0 [deg])

File: sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp [hunk 4]
```diff
@@ -96,6 +110,23 @@ void DualReturnOutlierFilterComponent::filter(
 
   uint32_t vertical_bins = vertical_bins_;
   uint32_t horizontal_bins = 36;
+  float max_azimuth = 36000.0f;
+  float min_azimuth = 0.0f;
+  switch (ROI_mode_) {
+    case 3: {
```
[the remark this answers is not in the document]
Comment (badai-nguyen) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp:116]: The variable is changed to lowercase, type of string.

File: sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp (hunk 2 above)
Comment (badai-nguyen) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp]: I agree. "pcl" is removed.

File: sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp [hunk 5]
```diff
@@ -96,6 +110,23 @@ void DualReturnOutlierFilterComponent::filter(
 
   uint32_t vertical_bins = vertical_bins_;
   uint32_t horizontal_bins = 36;
+  float max_azimuth = 36000.0f;
+  float min_azimuth = 0.0f;
+  switch (roi_mode_map_[roi_mode_]) {
+    case 3: {
+      max_azimuth = max_azimuth_deg_ * 100.0;
+      min_azimuth = min_azimuth_deg_ * 100.0;
```
[the remark this answers is not in the document]
Comment (badai-nguyen) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp:118]: The azimuth is times by 100 for directly comparing with azimuth of input point cloud. 

https://github.com/badai-nguyen/autoware.universe/blob/2a12a3bf0ed433377d80b63923e59f926acfae36/sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp#L273

https://github.com/badai-nguyen/autoware.universe/blob/2a12a3bf0ed433377d80b63923e59f926acfae36/sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp#L329

[the remark this answers is not in the document]
Comment (miursh) [reply sensing/pointcloud_preprocessor/src/outlier_filter/dual_return_outlier_filter_nodelet.cpp:114]: OK. I got it

File: sensing/pointcloud_preprocessor/docs/dual-return-outlier-filter.md [hunk 6]
```diff
@@ -47,6 +59,10 @@ This implementation inherits `pointcloud_preprocessor::Filter` class, please ref
...
 | `weak_first_local_noise_threshold` | int    | The parameter for determining whether it is noise                                                                         |
 | `visibility_threshold`             | float  | When the percentage of white pixels in the binary histogram falls below this parameter the diagnostic status becomes WARN |
+| `roi_mode`                         | string | The name of ROI mode for switching                                                                                        |
+| `min_azimuth_deg`                  | float  | The left limit of azimuth for `Fixed_azimuth_ROI` mode                                                                    |
+| `max_azimuth_deg`                  | float  | The right limit of azimuth for `Fixed_azimuth_ROI` mode                                                                   |
+| `max_distance`                     | float  | The limit distance for for `Fixed_azimuth_ROI` mode                                                                       |
```
[the remark this answers is not in the document]
Comment (miursh) [reply sensing/pointcloud_preprocessor/docs/dual-return-outlier-filter.md:71]: https://github.com/autowarefoundation/autoware.universe/pull/345/commits/6e0b6ceb8a48bcabf8442819f5459cadb72ab0c9

**Review by miursh — APPROVED** [review]
LGTM

**Review by miursh — DISMISSED** [review]

**Review by miursh — APPROVED** [review]

**Pull request merged**